src/main/java/io/cdap/plugin/pubnub/PubNubConfig.java

```diff
@@ -0,0 +1,467 @@
+                                                  proxyType));
+  }
+
+  /**
+   * @return a <code>String</code> type representing a cipher.
+   */
+  @Nullable
+  public String getCipherKey() {
+    return cipher;
+  }
+
+  /**
+   * @return a <code>Boolean</code> type specifying whether proxy is required.
+   */
+  @Nullable
+  public boolean hasProxy() {
+    return useProxy;
+  }
+
+  /**
+   * @return a <code>String</code> specifying the hostname of the proxy to be used to connect.
+   */
+  public String getProxyHostname() {
+    return proxyHostname;
+  }
+
+  /**
+   * @return a <code>Integer</code> specifying the proxy port.
+   */
+  public Integer getProxyPort() {
+    return proxyPort;
+  }
+
+  /**
+   * @return a instance of <code>PNReconnectPolicy</code> specifying how the client should backoff before reconnect.
+   */
+  public PNReconnectionPolicy getReconnectionPolicy() {
+    if (reconnectionPolicy.equalsIgnoreCase("none")) {
+      return PNReconnectionPolicy.NONE;
+    } else if (reconnectionPolicy.equalsIgnoreCase("linear")) {
+      return PNReconnectionPolicy.LINEAR;
+    } else if (reconnectionPolicy.equalsIgnoreCase("exponential")) {
+      return PNReconnectionPolicy.EXPONENTIAL;
+    }
+
+    throw new IllegalStateException(String.format("Unsupported value for reconnectionPolicy: '%s'",
+                                                  reconnectionPolicy));
+  }
+
+  /**
+   * @return a <code>Boolean</code> type specifying whether to use secure client connection to connect to PubNub.
+   */
+  @Nullable
+  public boolean isSecure() {
+    return secure;
+  }
+
+  /**
+   * @return a <code>Integer</code> type specifying subscribe timeout.
+   */
+  @Nullable
+  public Integer getSubscribeTimeout() {
+    return subscribeTimeout;
+  }
+
+  /**
+   * @return a <code>String</code> type specifying proxy username.
+   */
+  @Nullable
+  public String getProxyUsername() {
+    return proxyUsername;
+  }
+
+  /**
+   * @return a <code>String</code> type specifying proxy password.
+   */
+  @Nullable
+  public String getProxyPassword() {
+    return proxyPassword;
+  }
+
+  public void validate(FailureCollector failureCollector) {
+    IdUtils.validateReferenceName(referenceName, failureCollector);
+
+    try {
+      getReconnectionPolicy();
+    } catch (IllegalStateException ex) {
+      failureCollector.addFailure(ex.getMessage(), null)
+        .withConfigProperty(PROPERTY_RECONNECTION_POLICY);
+    }
+
+    if (hasProxy()) {
+      try {
+        if (getProxyType() == null) {
+          failureCollector.addFailure("Proxy enabled, but proxy type is specified as 'None'", null)
+            .withConfigProperty(PROPERTY_PROXY_TYPE);
+        }
+      } catch (IllegalStateException ex) {
+        failureCollector.addFailure(ex.getMessage(), null)
+          .withConfigProperty(PROPERTY_PROXY_TYPE);
+      }
+
+      if (getProxyHostname() == null || getProxyPort() == null) {
```

Comment:
this should be 2 separate checks. If hostname is given and port is not, the hostname should not be highlighted.

Reply:
Done.
I have also squashed the commits. Hopefully it's fine.